### PR TITLE
Enable >2GB models + allow model paths to be passed for generate_artifacts API

### PR DESCRIPTION
**Pull request by carzh**
### Description
Alternative design from #20942 

Allow users to pass in a model path for the generate_artifacts API. 

### Motivation and Context
- ONNX API calls such as the onnx checker + shape inference fail when given a model > 2GB, but work if a path to a model >2GB is passed in.

### CODE REVIEW

File: orttraining/orttraining/python/training/onnxblock/blocks.py [hunk 1]
```diff
@@ -47,7 +47,10 @@ def __call__(self, *args, **kwargs):
 
         output = self.build(*args, **kwargs)
 
-        onnx.checker.check_model(self.base, True)
+        if accessor._GLOBAL_ACCESSOR.has_path:
+            onnx.checker.check_model(accessor._GLOBAL_ACCESSOR.path, True)
```
Comment (skottmckay) [inline orttraining/orttraining/python/training/onnxblock/blocks.py:51]: Does the checker fail if the ModelProto is > 2GB and that's why we need to pass a path and let it load another copy using that?

File: orttraining/orttraining/python/training/onnxblock/loss/loss.py [hunk 2]
```diff
@@ -95,17 +95,17 @@ def build(self, scores_input_name: str, labels_name: str = "labels"):
             labels_input.type.tensor_type.elem_type = onnx.TensorProto.INT64
             # If the predictions are (num_examples x num_classes)
             # labels should be (num_examples,)
```
Comment (skottmckay) [inline orttraining/orttraining/python/training/onnxblock/loss/loss.py:97]: Is num_classes always the last dimension and we have a batch dimension or something like that making it > 2D and requiring usage of -1?

File: orttraining/orttraining/python/training/onnxblock/onnxblock.py [hunk 3]
```diff
@@ -70,7 +73,20 @@ def __call__(self, *args, **kwargs):
...
+
+            onnx.shape_inference.infer_shapes_path(TEMP_ONNX_PATH)
+            # shape inferenced model is saved to original path
+            self._model = onnx.load(TEMP_ONNX_PATH)
+
+            # clean-up temp files
+            if os.path.exists(TEMP_ONNX_PATH):
+                os.remove(TEMP_ONNX_PATH)
+            if os.path.exists(TEMP_ONNX_PATH):
+                os.remove(TEMP_ONNX_PATH)
```
Comment (skottmckay) [inline orttraining/orttraining/python/training/onnxblock/onnxblock.py:87]:
```suggestion
        with tempfile.TemporaryDirectory() as temp_dir:
            temp_model_path = temp_dir + "/model.onnx"
            onnx.save(self.base, temp_model_path, save_as_external_data=True, all_tensors_to_one_file=True)

            onnx.shape_inference.infer_shapes_path(temp_model_path)
            # shape inferenced model is saved to original path
            self._model = onnx.load(temp_model_path)
```

Probably not necessary to have a global for the temp filename, and tempfile.TemporaryDirectory can handle the cleanup for you.

File: orttraining/orttraining/python/training/onnxblock/onnxblock.py [hunk 4]
```diff
@@ -187,7 +203,21 @@ def __call__(self, *args, **kwargs):
 
         output = self.build(*args, **kwargs)
 
-        model = onnx.shape_inference.infer_shapes(accessor._GLOBAL_ACCESSOR.model)
+        model = None
+        if accessor._GLOBAL_ACCESSOR.has_path:
+            onnx.save(self.base, TEMP_ONNX_PATH, save_as_external_data=True, all_tensors_to_one_file=True)
```
Comment (skottmckay) [inline orttraining/orttraining/python/training/onnxblock/onnxblock.py:208]: Same suggestion as above

File: orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py [hunk 5]
```diff
@@ -318,6 +318,8 @@ def crossentropy_loss(prediction, target):
 
         ort_outs = ort_session.run(ort_output_names, ort_inputs)
         torch_outs = crossentropy_loss(pt_model(x), target)
+        print("ort_ outs", ort_outs)
```
Comment (skottmckay) [inline orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py:321]:
```suggestion
        print("ort outs", ort_outs)
```

File: orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py [hunk 6]
```diff
@@ -356,6 +358,8 @@ def bcewithlogits_loss(prediction, target):
 
         ort_outs = ort_session.run(ort_output_names, ort_inputs)
         torch_outs = bcewithlogits_loss(pt_model(x), target)
+        print("ort_ outs", ort_outs)
```
Comment (skottmckay) [inline orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py:361]:
```suggestion
        print("ort outs", ort_outs)
```

File: orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py [hunk 7]
```diff
@@ -1099,3 +1103,85 @@ def test_custom_optimizer_block():
...
+            requires_grad=requires_grad_params,
+            loss=artifacts.LossType.CrossEntropyLoss,
+            optimizer=artifacts.OptimType.AdamW,
+            artifact_directory=temp_dir,
+        )
+
+        assert os.path.exists(os.path.join(temp_dir, "training_model.onnx"))
+        assert os.path.exists(os.path.join(temp_dir, "eval_model.onnx"))
+        assert os.path.exists(os.path.join(temp_dir, "optimizer_model.onnx"))
+        assert os.path.exists(os.path.join(temp_dir, "checkpoint"))
```
Comment (skottmckay) [inline orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py:1131]: Does generate_artifacts not return an error or raise an exception if all these are not created?

File: orttraining/orttraining/python/training/onnxblock/loss/loss.py (hunk 2 above)
Comment (carzh) [reply orttraining/orttraining/python/training/onnxblock/loss/loss.py:97]: I changed it for cases like LLMs, where the predictions are (batch_size, sequence_length, vocab) & I made the assumption that this distribution would always be the last dimension. I previously ran into an issue with appending crossentropyloss with TinyLlama PoC and thought it would be an easy fix. I think the assumption is fine to make for now -- we can reconsider how the softmaxcrossentropy loss block looks like later down the line if there are models where num_classes is not the last dimension.

File: orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py (hunk 5 above)
Comment (carzh) [reply orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py]: oops 😅

File: orttraining/orttraining/python/training/onnxblock/loss/loss.py (hunk 2 above)
Comment (skottmckay) [reply orttraining/orttraining/python/training/onnxblock/loss/loss.py:97]: Could the comment be updated to capture this info?

File: orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py (hunk 7 above)
Comment (carzh) [reply orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py:1127]: Hmm, it usually emits an error if it's not successful, but I thought that I'd add the check just in case -- I can remove it if it's redundant though

File: orttraining/orttraining/python/training/onnxblock/blocks.py [hunk 8]
```diff
@@ -28,8 +29,9 @@ class Block(ABC):
         base (onnx.ModelProto): The base model that the subclass can manipulate.
     """
 
-    def __init__(self):
+    def __init__(self, temp_file_name="temp.onnx"):
         self.base = None
+        self.temp_onnx_path = temp_file_name
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/blocks.py:34]: Does the temp_file_name need to be known in the constructor? Can we just use `tempfile` python module and create a temporary directory on the go?

File: orttraining/orttraining/python/training/onnxblock/loss/loss.py (hunk 2 above)
Comment (carzh) [reply orttraining/orttraining/python/training/onnxblock/loss/loss.py]: yeah, good note -- pushed updated comment

File: orttraining/orttraining/python/training/onnxblock/onnxblock.py (hunk 3 above)
Comment (carzh) [reply orttraining/orttraining/python/training/onnxblock/onnxblock.py]: for visibility: the tempfile with TemporaryDirectory approach was causing a minor bug where ONNX was unable to find the external data file. 

Also, we realized that the ModelProto that is loaded and passed around does not contain the information from external data, only the metadata for where to find the external data, so the external data files generated by intermediary steps should be saved in some form.

Scott suggested an approach where we save to the original model path that's passed in, but in the case that the original model was saved with "all_tensors_to_one_file" set to True, subsequent save_model calls created a separate single data file named something different.

I'm going with the approach where the temp data files are retained until the very end of the generate_artifacts call then deleted, but open to other approaches / suggestions!

File: orttraining/orttraining/python/training/onnxblock/blocks.py [hunk 9]
```diff
@@ -47,10 +49,56 @@ def __call__(self, *args, **kwargs):
...
+
+            onnx.shape_inference.infer_shapes_path(self.temp_onnx_file_name)
+            # shape inferenced model is saved to original path
+            model = onnx.load(self.temp_onnx_file_name)
+
+            return model
+        else:
+            return onnx.shape_inference.infer_shapes(accessor._GLOBAL_ACCESSOR.model)
+
+    def release(self):
```
Comment (skottmckay) [inline orttraining/orttraining/python/training/onnxblock/blocks.py:93]: Having to manually call `release` can be problematic. e.g. if there's an exception in the usage code before the call to `release` it would get missed.

Could we use     
- add `def __del__(self):` cleanup in destructor
- or use contextlib.contextmanager to automate cleanup
  -  https://stackoverflow.com/questions/13379742/right-way-to-clean-up-a-temporary-folder-in-python-class

I would lean towards the `__del__` usage as it requires less of the user. not sure which is the more pythonic approach.

File: orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py (hunk 7 above)
Comment (skottmckay) [reply orttraining/orttraining/test/python/orttraining_test_ort_apis_onnxblock.py:1127]: If generate_artifacts raises an exception as expected, please add a comment to that effect before the asserts. 

```suggestion
        // generate_artifacts should have thrown if it didn't complete successfully. sanity check to validate all expected outputs were created.
        assert os.path.exists(os.path.join(temp_dir, "training_model.onnx"))
        assert os.path.exists(os.path.join(temp_dir, "eval_model.onnx"))
        assert os.path.exists(os.path.join(temp_dir, "optimizer_model.onnx"))
        assert os.path.exists(os.path.join(temp_dir, "checkpoint"))
```

Without that comment, code like this can be misleading. If I read this test without the comment I would assume I needed to add similar checks in my usage, spreading the misconception that generate_artifacts may leave things in an invalid state without warning.

**Review by baijumeswani — APPROVED** [review]

**Review by skottmckay — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]
